### PR TITLE
mungegithub: Fixes #422

### LINKED ISSUE #422: Possible for submit queue to merge a sha other than what was tested
There's a race condition in the submit queue where the queue may merge a sha other than what was tested:
For example: kubernetes/kubernetes#19716 : this PR was 5/5 checks green, then needed rebase after the merge bot had called for a final test.  I rebased, repushed, and re-added the LGTM label, and the PR merged even though 1/5 checks was green (presumably CLA check) in github for that sha.

@eparis

### PR DESCRIPTION
**Pull request by apelisse**
At least attempts to fix one or two of the problems in #422.

Refresh the pull-request after we've done the tests to:
- Make sure it is still mergeable (doesn't have do-no-merge label or
still has lgtm label)
- The SHA hasn't changed, so that we don't commit something we haven't
tested

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/2008)

### CODE REVIEW

File: mungegithub/mungers/submit-queue.go [hunk 1]
```diff
@@ -1165,7 +1172,32 @@ func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
...
+		if err != nil {
+			glog.Errorf("%d: unknown err: %v", *obj.Issue.Number, err)
+			sq.SetMergeStatus(obj, unknown)
+			return true
+		}
+	}
+
+	// We shouldn't merge if it's not valid anymore
+	if !sq.validForMerge(obj) {
+		glog.Errorf("%d: Not mergeable anymore. Do not merge.", *obj.Issue.Number)
```
Comment (rmmh) [inline mungegithub/mungers/submit-queue.go]: we should probably do sq.setMergeStatus here with the "sha changed", and return false to do a retest.

File: mungegithub/mungers/submit-queue.go [hunk 2]
```diff
@@ -1165,7 +1172,32 @@ func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
...
+	}
+
+	if newSha, _, ok := obj.GetHeadAndBase(); !ok {
+		glog.Errorf("%d: Unable to get SHA", *obj.Issue.Number)
+		sq.SetMergeStatus(obj, unknown)
+		return true
+	} else if newSha != sha {
+		glog.Errorf("%d: Changed while running the test. Do not merge.", *obj.Issue.Number)
+		sq.SetMergeStatus(obj, headCommitChanged)
 		return true
```
Comment (rmmh) [inline mungegithub/mungers/submit-queue.go]: this should return false to do a retest, maybe?

File: mungegithub/mungers/submit-queue.go (hunk 1 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: `sq.validForMerge()` does it for us

File: mungegithub/mungers/submit-queue.go (hunk 2 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: Good point

**Comment by apelisse** [comment]
Updated @rmmh

**Comment by rmmh** [comment]
/lgtm

Do the unit tests exercise this path? How disgusting would it be to test this functionality?

**Comment by apelisse** [comment]
Yes, unit-tests are going there (I also run them successfully). I agree it'd be nice to be able to test that more thorougly.

**Comment by apelisse** [comment]
Replaced tabs with space for alignment. Auto-lgtming ...

**Comment by apelisse** [comment]
No, but I can add the label myself

File: mungegithub/mungers/submit-queue.go (hunk 1 above)
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: Why is this a glog.Errorf() ?

File: mungegithub/mungers/submit-queue.go [hunk 3]
```diff
@@ -1165,10 +1172,35 @@ func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
...
+	}
+
+	// We shouldn't merge if it's not valid anymore
+	if !sq.validForMerge(obj) {
+		glog.Errorf("%d: Not mergeable anymore. Do not merge.", *obj.Issue.Number)
 		return true
 	}
 
+	if newSha, _, ok := obj.GetHeadAndBase(); !ok {
+		glog.Errorf("%d: Unable to get SHA", *obj.Issue.Number)
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: Do we need to spam the logs? shouldn't we have already done that in github.go?

File: mungegithub/mungers/submit-queue.go [hunk 4]
```diff
@@ -1165,10 +1172,35 @@ func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
...
+		sq.SetMergeStatus(obj, unknown)
+		return true
+	} else if newSha != sha {
+		glog.Errorf("%d: Changed while running the test. Do not merge.", *obj.Issue.Number)
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: Maybe this belongs in the logs, but its certainly not an Errorf...

File: mungegithub/mungers/submit-queue.go [hunk 5]
```diff
@@ -1165,10 +1172,35 @@ func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
...
+		sq.SetMergeStatus(obj, headCommitChanged)
+		return false
+	}
+
 	if !sq.e2eStable(true) {
 		if sq.validForMerge(obj) {
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: does this still serve a purpose?

**Review by eparis — COMMENTED** [review]
Anyone have time to clean this up?

**Comment by eparis** [comment]
Also, this doesn't actually fix the problem...  It makes the race smaller, but it isn't fixed...

fixing it requires extending go-github. I actually have an open PR for that...  I'll try to get back to it next week.

I also didn't see an explanation why there was no unit test added for this case. It's certainly possible to do so...

**Comment by apelisse** [comment]
I should have paid more attention to log verbosity. I will address that now.

Adding link to pull-request for the record: https://github.com/google/go-github/pull/456

OK So I will remove the part that checks that the sha hasn't changed, but we should probably keep the refresh and check for mergeability (we should still not merge if the lgtm is gone, or do-not-merge label has appeared).

File: mungegithub/mungers/submit-queue.go (hunk 3 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: Removing this

File: mungegithub/mungers/submit-queue.go (hunk 4 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: Removing this

File: mungegithub/mungers/submit-queue.go (hunk 5 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: removed the condition

**Pull request merged**